### DISCUSSION TITLE
Cast call function returning empty string hex value

### DISCUSSION THREAD

**Discussion by shadowalpha197** [Cyfrin/foundry-full-course-cu#1706] [Q&A]
In the simplestorage.sol contract I passed the value 45 in the store(uint256) function from the vs code terminal, but when I called the function retrieve() in the terminal, it gave the value '0x', even though it's not supposed too(0x is basically no hex value). What should I do? Thanks.

**Comment by EngrPips** [comment]
Hello @shadowalpha197, I ran into the same problem recently, and I spent over 2 hours trying to understand where the problem lies, but no luck at all. I would advice you check Foundry Github page https://github.com/foundry-rs/foundry to see if someone have reported similar problem and what the suggested solution and if no one had initially reported it maybe you should report and look forward to people helping.

**Comment by Ndife** [comment]
> In the simplestorage.sol contract I passed the value 45 in the store(uint256) function from the vs code terminal, but when I called the function retrieve() in the terminal, it gave the value '0x', even though it's not supposed too(0x is basically no hex value). What should I do? Thanks.

Hello @shadowalpha197,

I encountered the same issue recently when calling the `retrieve() `function after storing a value using `store(uint256)`. Have you been able to resolve this error?

**Reply by EngrPips** [reply]
Did you broadcast the transaction when you deployed the `SimpleStorage` contract.

**Reply by Ndife** [reply]
Yes, I did. 

`forge script script/DeploySimpleStorage.s.sol --rpc-url $RPC_URL --broadcast --account defaultKey`

**Reply by EngrPips** [reply]
and what was the output of that?

**Reply by Ndife** [reply]
Here's the output: 

`[⠊] Compiling...
No files changed, compilation skipped
Enter keystore password:
Script ran successfully.

== Return ==
0: contract SimpleStorage 0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9

## Setting up 1 EVM.

==========================

Chain 31337

Estimated gas price: 1.536845937 gwei

Estimated total gas used for script: 471042

Estimated amount required: 0.000723918983856354 ETH

==========================

##### anvil-hardhat
✅  [Success]Hash: 0x00b1cbfeaa59dd9da960391bc373639e53cbcac2a757bb8e072dcb87409fff4e
Contract Address: 0xCf7Ed3AccA5a467e9e704C703E8D87F634fB0Fc9
Block: 4
Paid: 0.000243737650737728 ETH (362432 gas * 0.672505879 gwei)

✅ Sequence #1 on anvil-hardhat | Total Paid: 0.000243737650737728 ETH (362432 gas * avg 0.672505879 gwei)
                                                                                                  

==========================

ONCHAIN EXECUTION COMPLETE & SUCCESSFUL.

Transactions saved to: /Users/ndifeuchenna/Documents/solidarity/foundry/foundry-simple-storage-f24/broadcast/DeploySimpleStorage.s.sol/31337/run-latest.json

Sensitive values saved to: /Users/ndifeuchenna/Documents/solidarity/foundry/foundry-simple-storage-f24/cache/DeploySimpleStorage.s.sol/31337/run-latest.json
`

**Reply by EngrPips** [reply]
Now run the command to store the value and then retrieve it, let's see what will the outcome.

**Reply by EngrPips** [reply]
The below is my whole process of deploying the contract, storing value in the contract and retrieving value from the contract. Everything works smoothly for me.

```js
loyalty@Loyaltys-MacBook-Pro FOUNDRYSTORAGE % forge script script/DeploySimpleStorage.s.sol --rpc-url $RPC_URL --broadcast --account defaultKey
[⠒] Compiling...
No files changed, compilation skipped
Enter keystore password:
Script ran successfully.

== Return ==
0: contract SimpleStorage 0x5FbDB2315678afecb367f032d93F642f64180aa3

## Setting up 1 EVM.

==========================

Chain 31337

Estimated gas price: 2.000000001 gwei

Estimated total gas used for script: 455465

Estimated amount required: 0.000910930000455465 ETH

==========================

##### anvil-hardhat
✅  [Success]Hash: 0xea39a2678a611d47047e2301804ddf1477a1ba2e81b285d203b7151fa52d1b73
Contract Address: 0x5FbDB2315678afecb367f032d93F642f64180aa3
Block: 1
Paid: 0.000350358000350358 ETH (350358 gas * 1.000000001 gwei)

✅ Sequence #1 on anvil-hardhat | Total Paid: 0.000350358000350358 ETH (350358 gas * avg 1.000000001 gwei)
                                                                                                                                                                        

==========================

ONCHAIN EXECUTION COMPLETE & SUCCESSFUL.

Transactions saved to: /Users/loyalty/Desktop/CyfrinGig/FOUNDRYFUNDAMENTALS/FOUNDRYSTORAGE/broadcast/DeploySimpleStorage.s.sol/31337/run-latest.json

Sensitive values saved to: /Users/loyalty/Desktop/CyfrinGig/FOUNDRYFUNDAMENTALS/FOUNDRYSTORAGE/cache/DeploySimpleStorage.s.sol/31337/run-latest.json

loyalty@Loyaltys-MacBook-Pro FOUNDRYSTORAGE % cast send 0x5FbDB2315678afecb367f032d93F642f64180aa3 "store(uint256)" 7 --rpc-url $RPC_URL --account defaultKey
Enter keystore password:

blockHash               0xd5ed7dd53974a2fad85191e33fe8916b36dea08c5e9105d380a46ccb996f252d
blockNumber             2
contractAddress         
cumulativeGasUsed       43535
effectiveGasPrice       877919651
from                    0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
gasUsed                 43535
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
root                    0x24ed2885833623091879e93408e2019bc3bd3f7e07c6829398f615acae80033c
status                  1 (success)
transactionHash         0x9e5fabb3b5e1238211bcdbff9c7486dcd782faf5bc74a7796137f1cf2c5cb290
transactionIndex        0
type                    2
blobGasPrice            1
blobGasUsed             
authorizationList       
to                      0x5FbDB2315678afecb367f032d93F642f64180aa3

cast call 0x5FbDB2315678afecb367f032d93F642f64180aa3 "retrieve()" --rpc-url $RPC_URL
0x0000000000000000000000000000000000000000000000000000000000000007


```

**Reply by Ndife** [reply]
ok. but i'm still having the same error. below is my whole process and the output


```

forge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account defforge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account defaultKey
[⠊] Compiling...
No files changed, compilation skipped
Enter keystore password:
Script ran successfully.

== Return ==
0: contract SimpleStorage 0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9

## Setting up 1 EVM.

==========================

Chain 31337

Estimated gas price: 1.353766049 gwei

Estimated total gas used for script: 471042

Estimated amount required: 0.000637680667253058 ETH

==========================

##### anvil-hardhat
✅  [Success]Hash: 0x8ec4946c661ed6c5affce565569bdf6c57a22e3df99f4abc3225b4174f581f0b
Contract Address: 0xDc64a140Aa3E981100a9becA4E685f962f0cF6C9
Block: 5
Paid: 0.000214701909039296 ETH (362432 gas * 0.592392253 gwei)

✅ Sequence #1 on anvil-hardhat | Total Paid: 0.000214701909039296 ETH (362432 gas * avg 0.592392253 gwei)
                                                                                                                                     

==========================

ONCHAIN EXECUTION COMPLETE & SUCCESSFUL.

Transactions saved to: /Users/ndifeuchenna/Documents/solidity/foundry/foundry-simple-storage-f24/broadcast/DeploySimpleStorage.s.sol/31337/run-latest.json

Sensitive values saved to: /Users/ndifeuchenna/Documents/solidity/foundry/foundry-simple-storage-f24/cache/DeploySimpleStorage.s.sol/31337/run-latest.json

ndifeuchenna@Ndifes-MBP foundry-simple-storage-f24 % cast send 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 "store(uint256)" 7 --rpc-ur
l http://127.0.0.1:8545 --account defaultKey

Enter keystore password:

blockHash               0x8a3778d2cb76d6e18735bcfca9efd80fdb8e63275760d9fc5b01fb3f64192e7d
blockNumber             6
contractAddress         
cumulativeGasUsed       21204
effectiveGasPrice       520132405
from                    0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
gasUsed                 21204
logs                    []
logsBloom               0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
root                    0xc942d3d86ee2556ec7ad5f580822a64cfe48b32901b2766205d257ed269f5fdf
status                  1 (success)
transactionHash         0xe0fa39c40b8cc9ce88e7851b190c43acd056318f5daa27e5309cf912bd78acd8
transactionIndex        0
type                    2
blobGasPrice            1
blobGasUsed             
authorizationList       
to                      0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266
ndifeuchenna@Ndifes-MBP foundry-simple-storage-f24 % cast call 0xf39fd6e51aad88f6f4ce6ab8827279cfffb92266 "retrieve()"  --rpc-url http://127.0.0.1:8545                                                                      "retrieve()"  --rpc-url http://127.0.0.1:8545
0x


```

**Reply by EngrPips** [reply]
run `foundryup` then retry.

**Reply by Ndife** [reply]
I just did. it's still the same result `0x`

**Reply by EngrPips** [reply]
Interesting, I am unsure what is wrong on your end, but maybe check the official foundry repo https://github.com/foundry-rs/foundry and see if anyone has reported the same issue lately.

**Reply by Ndife** [reply]
ok. Thank you

**Reply by EngrPips** [reply]
You are welcome.

**Reply by Ndife** [reply]
It's working now. I was using the wrong contract address. Instead of the one generated from the deployment logs, I mistakenly used the address created when I ran this command:
cast wallet import defaultKey --interactive
(This address corresponds to the wallet generated during the encryption of the private key.)

After switching to the correct contract address from the deployment logs, the contract functions worked as expected.

**Reply by EngrPips** [reply]
Interesting

**Comment by Ndife** [comment]
@shadowalpha197 make sure that you're using the contract address generated from the deployment logs, when you run this command 

`forge script script/DeploySimpleStorage.s.sol --rpc-url http://127.0.0.1:8545 --broadcast --account defaultKey`.

and not the one generated during the encryption of the private key.
